aocdbc/day_4_passport_processing.py

```diff
@@ -0,0 +1,107 @@
+from re import fullmatch
+import sys
+from typing import Dict, List, Optional
+
+from icontract import require, ensure
+
+_REQUIRED_KEYS = set(["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"])
+
+
+@ensure(lambda: True)
+def blank_line_split(text: str) -> List[str]:
+    # NOTE: I tried to do a lookahead for the 2nd newline, but split doesn't take
+    # a regex.
+    return text.split("\n\n")
+
+
+@ensure(lambda: True)
+def parse(passport: str) -> Optional[Dict[str, str]]:
+    result = {}
+    for tok in passport.split():
+        pair = tok.split(":", 2)
+        if len(pair) != 2:
+            return None
+        result[pair[0]] = pair[1]
+    return result
+
+
+@ensure(lambda entry, result: result == (not (_REQUIRED_KEYS - entry.keys())))
+def is_valid(entry: Dict[str, str]) -> bool:
+    return len(_REQUIRED_KEYS - entry.keys()) == 0
+
+
+@ensure(lambda result: result >= 0)
+def count_valid(batch: str) -> int:
```

Comment:
Assuming `count_valid(lines: Lines), additional requirement for `count_valid` could be :
```
@require(
	lambda lines:
	all(
		len(line) == 0 or 
		PASSPORT_RE.match(line)
	)
)
```

Reply:
Any thoughts about dealing with non-duplicated keys here?
Right now, I'm having trouble imaging how this isn't just a game of duplicating logic between contracts and implementations.

Reply:
I hope I got the question right (otherwise maybe we can have a quick call this week). The logic is not duplicated directly, but it's similar. Here's a sketch:
```python
def parse_entries(passport_text) -> List[Tuple[str, str]]:
    ...

@require(lambda entries: (keys := [entry[0] for entry in entries], len(set(keys)) == len(keys))
def interpret_entries(entries: List[Tuple[str, str]]) -> MutableMapping[str, str]:
    return {key: value for key, value in entries}
```

This is indeed a bit artificial, as the code is so simple, but it's a nice showcase as well as an additional example for testing the tools.

Reply:
Hmm, so this is how I'd write it (the dictionary constructor can already take an iterable of pairs):
```
@require(lambda entries: len(dict(entries)) == len(entries))
def parse_passport(entries: List[Tuple[str, str]]) -> Dict[str, str]:
    return dict(entries)
```

Feels strange to write a function for this though. My original question was about the requirements that such a precondition places on `count_valid` - your  original suggested precondition for `count_valid` isn't enough to guarantee that keys aren't duplicated.

Reply:
> My original question was about the requirements that such a precondition places on count_valid - your original suggested precondition for count_valid isn't enough to guarantee that keys aren't duplicated.

Ah, sorry, now I understand.

You are right. I don't think there is a meaningful set of pre-conditions -- other than re-implementing everything -- on `count_valid`. I suppose you can only test the individual parts meaningfully in an automatic way.

> Feels strange to write a function for this though.

(After a second pass over the code) I agree. What do you think about the following idea:
* (assumption) You don't the dictionary in the public interface at all.
* Make `is_valid` operate on list of entries.
* `is_valid` checks also for unique entries.
* `parse_passport` goes away and is integrated within `is_valid`.
* The `count_valid` is then:
  ```
  passport_texts = blank_line_split(batch)
  return sum(
    1
    for passport_text in blank_line_split(batch)
    is_valid(entries=parse_passport_entries(passport_text))
  )
  ```

There will be a meaningful exception if something goes wrong. What do you think?